In [93]:
# load the libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
output_width = 1000
#output_width = 80 #//*** Normal Output width
pd.set_option("display.width", output_width)
pd.set_option('display.max_rows', 100)
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)

In [59]:
# Load the final data into a dataframe
housing_modeling_org = pd.read_csv("housing_modeling_org.csv",sep=",")

In [60]:
# display the dataframe
housing_modeling_org

Unnamed: 0  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1 BsmtFinType2 HeatingQC CentralAir  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu GarageType GarageFinish  GarageCars  GarageArea GarageCond PavedDrive  PoolArea  MoSold  YrSold SaleCondition  TotalSF  TotalBath  TotalPorchSF  TotalBathSum  SalePrice
0              0          60       RL         65.0     8450   Pave      Reg         Lvl    Inside       Gtl      CollgCr       Norm       Norm     1Fam     2Story            7            5       2003          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace    BrkFace        Gd        TA      PConc       Gd       TA           No          GLQ          Unf        Ex          Y             3             1          Gd             8        Typ           0          Gd     Attchd          RFn           2         548         TA          Y         0       2    2008        Normal     2566          3            61           3.5     208500
1              1          20       RL         80.0     9600   Pave      Reg         Lvl       FR2       Gtl      Veenker      Feedr       Norm     1Fam     1Story            6            8       1976          1976     Gable  CompShg     MetalSd     MetalSd       None       None        TA        TA     CBlock       Gd       TA           Gd          ALQ          Unf        Ex          Y             3             1          TA             6        Typ           1          TA     Attchd          RFn           2         460         TA          Y         0       5    2007        Normal     2524          2             0           2.5     181500
2              2          60       RL         68.0    11250   Pave      IR1         Lvl    Inside       Gtl      CollgCr       Norm       Norm     1Fam     2Story            7            5       2001          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace    BrkFace        Gd        TA      PConc       Gd       TA           Mn          GLQ          Unf        Ex          Y             3             1          Gd             6        Typ           1          TA     Attchd          RFn           2         608         TA          Y         0       9    2008        Normal     2706          3            42           3.5     223500
3              3          70       RL         60.0     9550   Pave      IR1         Lvl    Corner       Gtl      Crawfor       Norm       Norm     1Fam     2Story            7            5       1915          1970     Gable  CompShg     Wd Sdng     Wd Shng       None       None        TA        TA     BrkTil       TA       Gd           No          ALQ          Unf        Gd          Y             3             1          Gd             7        Typ           1          Gd     Detchd          Unf           3         642         TA          Y         0       2    2006       Abnorml     2473          1           307           2.0     140000
4              4          60       RL         84.0    14260   Pave      IR1         Lvl       FR2       Gtl      NoRidge       Norm       Norm     1Fam     2Story            8            5       2000          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace    BrkFace        Gd        TA      PConc       Gd       TA           Av          GLQ          Unf        Ex          Y             4             1          Gd             9        Typ           1          TA     Attchd          RFn           3         836         TA          Y         0      12    2008        Normal     3343          3            84           3.5     250000
...          ...         ...      ...          ...      ...    ...      ...         ...       ...       ...          ...        ..

In [61]:
# viewing the column info
housing_modeling_org.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 57 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1460 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   LotShape       1460 non-null   object 
 7   LandContour    1460 non-null   object 
 8   LotConfig      1460 non-null   object 
 9   LandSlope      1460 non-null   object 
 10  Neighborhood   1460 non-null   object 
 11  Condition1     1460 non-null   object 
 12  Condition2     1460 non-null   object 
 13  BldgType       1460 non-null   object 
 14  HouseStyle     1460 non-null   object 
 15  OverallQual    1460 non-null   int64  
 16  OverallCond    1460 non-null   int64  
 17  YearBuilt      1460 non-null   int64  
 18  YearRemo

In [62]:
# Below is the list of columns to be dropped
cols_to_drop=['MSSubClass','MSZoning','Street','LotShape','LandContour','LotConfig','LandSlope','Neighborhood','Condition1','Condition2','OverallQual','YearRemodAdd','TotalBath','TotalPorchSF'
             ,'SaleCondition','MoSold','PavedDrive','GarageCond','GarageFinish','GarageType','FireplaceQu','KitchenQual','BsmtFinType1','BsmtFinType2',
             'BsmtExposure','BsmtCond','BsmtQual','Exterior1st','Exterior2nd','ExterQual','ExterCond','MasVnrType','MasVnrArea']
# Making a copy of the original dataframe to be worked on
housing_modeling_v1 = housing_modeling_org
print("Dimension before drop:", housing_modeling_org.shape)
housing_modeling_v1 = housing_modeling_org.drop(cols_to_drop,axis=1)
print("Dimension after drop:", housing_modeling_v1.shape)

Dimension before drop: (1460, 57)
Dimension after drop: (1460, 24)


In [63]:
# Saving the feature set till now
housing_modeling_v1.to_csv("housing_modeling_v1.csv")

In [64]:
# store columns with specific data type
integer_columns = housing_modeling_v1.select_dtypes(include=['int64']).columns
float_columns = housing_modeling_v1.select_dtypes(include=['float64']).columns
object_columns = housing_modeling_v1.select_dtypes(include=['object']).columns
  
# display columns
print('\nint64 columns:\n',integer_columns)
print('\nfloat64 columns:\n',float_columns)
print('\nobject columns:\n',object_columns)


int64 columns:
 Index(['Unnamed: 0', 'LotArea', 'OverallCond', 'YearBuilt', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces', 'GarageCars', 'GarageArea', 'PoolArea', 'YrSold', 'TotalSF', 'SalePrice'], dtype='object')

float64 columns:
 Index(['LotFrontage', 'TotalBathSum'], dtype='object')

object columns:
 Index(['BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Foundation', 'HeatingQC', 'CentralAir', 'Functional'], dtype='object')


In [65]:
# Separating out the object columns into a different dataframe
housing_modeling_v1_object = housing_modeling_v1.loc[:,object_columns]

In [67]:
housing_modeling_v1_object.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   BldgType    1460 non-null   object
 1   HouseStyle  1460 non-null   object
 2   RoofStyle   1460 non-null   object
 3   RoofMatl    1460 non-null   object
 4   Foundation  1460 non-null   object
 5   HeatingQC   1460 non-null   object
 6   CentralAir  1460 non-null   object
 7   Functional  1460 non-null   object
dtypes: object(8)
memory usage: 91.4+ KB


In [68]:
#applying label encoder to the entire dataframe
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

# This will convert only the categorical columns in the entire dataframe
# all categorical columns are in the list object "object_columns"
# The following for loop will convert only if the column is in the object list
for col in housing_modeling_v1[object_columns]:
    housing_modeling_v1[col]=le.fit_transform(housing_modeling_v1[col])

In [69]:
housing_modeling_v1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 24 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    1460 non-null   int64  
 1   LotFrontage   1460 non-null   float64
 2   LotArea       1460 non-null   int64  
 3   BldgType      1460 non-null   int64  
 4   HouseStyle    1460 non-null   int64  
 5   OverallCond   1460 non-null   int64  
 6   YearBuilt     1460 non-null   int64  
 7   RoofStyle     1460 non-null   int64  
 8   RoofMatl      1460 non-null   int64  
 9   Foundation    1460 non-null   int64  
 10  HeatingQC     1460 non-null   int64  
 11  CentralAir    1460 non-null   int64  
 12  BedroomAbvGr  1460 non-null   int64  
 13  KitchenAbvGr  1460 non-null   int64  
 14  TotRmsAbvGrd  1460 non-null   int64  
 15  Functional    1460 non-null   int64  
 16  Fireplaces    1460 non-null   int64  
 17  GarageCars    1460 non-null   int64  
 18  GarageArea    1460 non-null 

In [71]:
# viewing the column info
housing_modeling_v1.head(10)

Unnamed: 0  LotFrontage  LotArea  BldgType  HouseStyle  OverallCond  YearBuilt  RoofStyle  RoofMatl  Foundation  HeatingQC  CentralAir  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd  Functional  Fireplaces  GarageCars  GarageArea  PoolArea  YrSold  TotalSF  TotalBathSum  SalePrice
0           0         65.0     8450         0           5            5       2003          1         1           2          0           1             3             1             8           6           0           2         548         0    2008     2566           3.5     208500
1           1         80.0     9600         0           2            8       1976          1         1           1          0           1             3             1             6           6           1           2         460         0    2007     2524           2.5     181500
2           2         68.0    11250         0           5            5       2001          1         1           2          0           1             3             1             6           6           1           2         608         0    2008     2706           3.5     223500
3           3         60.0     9550         0           5            5       1915          1         1           0          2           1             3             1             7           6           1           3         642         0    2006     2473           2.0     140000
4           4         84.0    14260         0           5            5       2000          1         1           2          0           1             4             1             9           6           1           3         836         0    2008     3343           3.5     250000
5           5         85.0    14115         0           0            5       1993          1         1           5          0           1             1             1             5           6           0           2         480         0    2009     2158           2.5     143000
6           6         75.0    10084         0           2            5       2004          1         1           2          0           1             3             1             7           6           1           2         636         0    2007     3380           3.0     307000
7           7         60.0    10382         0           5            6       1973          1         1           1          0           1             3             1             7           6           2           2         484         0    2009     3197           3.5     200000
8           8         51.0     6120         0           0            5       1931          1         1           0          2           1             2             2             8           2           2           2         468         0    2008     2726           2.0     129900
9           9         50.0     7420         1           1            6       1939          1         1           0          0           1             2             2             5           6           2           1         205         0    2008     2068           2.0     118000

In [72]:
housing_modeling_v1["AgeOfHouseSold"] = (
    housing_modeling_v1["YrSold"] - housing_modeling_v1["YearBuilt"]
)

In [77]:
# viewing the column info
housing_modeling_v1.head(10)

Unnamed: 0  LotFrontage  LotArea  BldgType  HouseStyle  OverallCond  RoofStyle  RoofMatl  Foundation  HeatingQC  CentralAir  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd  Functional  Fireplaces  GarageCars  GarageArea  PoolArea  TotalSF  TotalBathSum  SalePrice  AgeOfHouseSold
0           0         65.0     8450         0           5            5          1         1           2          0           1             3             1             8           6           0           2         548         0     2566           3.5     208500               5
1           1         80.0     9600         0           2            8          1         1           1          0           1             3             1             6           6           1           2         460         0     2524           2.5     181500              31
2           2         68.0    11250         0           5            5          1         1           2          0           1             3             1             6           6           1           2         608         0     2706           3.5     223500               7
3           3         60.0     9550         0           5            5          1         1           0          2           1             3             1             7           6           1           3         642         0     2473           2.0     140000              91
4           4         84.0    14260         0           5            5          1         1           2          0           1             4             1             9           6           1           3         836         0     3343           3.5     250000               8
5           5         85.0    14115         0           0            5          1         1           5          0           1             1             1             5           6           0           2         480         0     2158           2.5     143000              16
6           6         75.0    10084         0           2            5          1         1           2          0           1             3             1             7           6           1           2         636         0     3380           3.0     307000               3
7           7         60.0    10382         0           5            6          1         1           1          0           1             3             1             7           6           2           2         484         0     3197           3.5     200000              36
8           8         51.0     6120         0           0            5          1         1           0          2           1             2             2             8           2           2           2         468         0     2726           2.0     129900              77
9           9         50.0     7420         1           1            6          1         1           0          0           1             2             2             5           6           2           1         205         0     2068           2.0     118000              69

In [76]:
housing_modeling_v1 = housing_modeling_v1.drop(columns=['YearBuilt','YrSold'])

In [78]:
# Saving the feature set till now
housing_modeling_v1.to_csv("housing_modeling_v1_num.csv")

In [79]:
housing_modeling_v1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      1460 non-null   int64  
 1   LotFrontage     1460 non-null   float64
 2   LotArea         1460 non-null   int64  
 3   BldgType        1460 non-null   int64  
 4   HouseStyle      1460 non-null   int64  
 5   OverallCond     1460 non-null   int64  
 6   RoofStyle       1460 non-null   int64  
 7   RoofMatl        1460 non-null   int64  
 8   Foundation      1460 non-null   int64  
 9   HeatingQC       1460 non-null   int64  
 10  CentralAir      1460 non-null   int64  
 11  BedroomAbvGr    1460 non-null   int64  
 12  KitchenAbvGr    1460 non-null   int64  
 13  TotRmsAbvGrd    1460 non-null   int64  
 14  Functional      1460 non-null   int64  
 15  Fireplaces      1460 non-null   int64  
 16  GarageCars      1460 non-null   int64  
 17  GarageArea      1460 non-null   i

In [80]:
# Load Data
# Load the final data into a dataframe
df_lr = pd.read_csv("housing_modeling_v1_num.csv",sep=",")
features = df_lr.drop(columns="SalePrice").values
target = df_lr["SalePrice"].values

In [107]:
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Lasso
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score using 

In [88]:
scaler = StandardScaler()
feature_scaled = scaler.fit_transform(features)

In [96]:
# Creating the training and test data sets
X = feature_scaled
y = target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=40)
print(X_train.shape); print(X_test.shape)

(1022, 23)
(438, 23)


In [97]:
# Linear Regression
lr = LinearRegression()
lr.fit(X_train, y_train)

LinearRegression()

In [102]:
pred_train_lr= lr.predict(X_train)
print(np.sqrt(mean_squared_error(y_train,pred_train_lr)))
print(r2_score(y_train, pred_train_lr))

pred_test_lr= lr.predict(X_test)
print(np.sqrt(mean_squared_error(y_test,pred_test_lr))) 
print(r2_score(y_test, pred_test_lr))

39581.055516358814
0.7688617716204492
31031.708739538364
0.8150505070548879


f "The above output shows that the RMSE, one of the two evaluation metrics, is" {39581} for train data and {31031} for test data. On the other hand, R-squared value is 77 percent for train data and 82 percent for test data, which is a good performance.

Lasso regression, or the Least Absolute Shrinkage and Selection Operator, is also a modification of linear regression. In Lasso, the loss function is modified to minimize the complexity of the model by limiting the sum of the absolute values of the model coefficients (also called the l1-norm).

In [105]:
model_lasso = Lasso(alpha=0.01)
model_lasso.fit(X_train, y_train) 
pred_train_lasso= model_lasso.predict(X_train)
print(np.sqrt(mean_squared_error(y_train,pred_train_lasso)))
print(r2_score(y_train, pred_train_lasso))

pred_test_lasso= model_lasso.predict(X_test)
print(np.sqrt(mean_squared_error(y_test,pred_test_lasso))) 
print(r2_score(y_test, pred_test_lasso))

39581.0555163909
0.7688617716200744
31031.70185859434
0.8150505890759573


In [108]:
#RF model
model_rf = RandomForestRegressor(n_estimators=500, oob_score=True, random_state=100)
model_rf.fit(X_train, y_train) 
pred_train_rf= model_rf.predict(X_train)
print(np.sqrt(mean_squared_error(y_train,pred_train_rf)))
print(r2_score(y_train, pred_train_rf))

pred_test_rf = model_rf.predict(X_test)
print(np.sqrt(mean_squared_error(y_test,pred_test_rf)))
print(r2_score(y_test, pred_test_rf))

13210.777625859055
0.9742513892592122
27568.866346838502
0.8540246315336022


The above output shows that the RMSE and R-squared values on the training data is 13210 and 97.4 percent, respectively. For the test data, the result for these metrics is 27,568 and 85.4 percent, respectively. The performance of the random forest model is far superior.